spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala

```diff
@@ -55,11 +55,19 @@ class Adagrad[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeri
     val lrd = _config.get[Double]("learningRateDecay").getOrElse(0.0)
     val nevals = _state.get[Int]("evalCounter").getOrElse(0)
     val wd = config.get[Double]("weightDecay").getOrElse(0.0)
+    val l1wd = config.get[Double]("l1WeightDecay").getOrElse(0.0)
+    val l2wd = config.get[Double]("l2WeightDecay").getOrElse(wd)
+    val weightsSignBuffer = _state.get[Tensor[T]]("weightSigns").getOrElse(Tensor(parameter.size()))
```

Comment:
What's the weightsSignBuffer for? Do we always need it? It is bad for perf if the parameter is very large.